### ISSUE TITLE
zsh: ~/.profile is sourced twice

### ISSUE THREAD

**Issue by Minnozz** [m-ou-se/config#15]
Once from `zshenv` (for login/interactive/non-interactive shells) and once from `zshrc` (for login/interactive shells).

**Comment by m-ou-se** [comment]
I vaguely remember this is on purpose, because `zshenv` was not loaded in all cases or something like that. (`.profile` is also idempotent on purpose.)

@de-vri-es do you remember why both `zshrc` and `zenv` load `.profile`?

**Comment by m-ou-se** [comment]
Ah, looks like I did that. Explanation in commit 61362aefc88dc5fb2b82b12a66191b4eef457265:

```
commit 61362aefc88dc5fb2b82b12a66191b4eef457265
Author: Mara Bos <m-ou.se@m-ou.se>
Date:   Sat Jul 8 15:58:22 2017 +0200

    Source .profile in zshrc as well.
    
    For login shells, zsh loads /etc/profile after zshenv, which will
    overwrite $PATH. zshrc will be loaded afterwards, so it can correct it.
    zshenv still needs to source .profile, for non-interactive shells.
```

**Comment by m-ou-se** [comment]
Until 3 years ago, the default Arch Linux `/etc/profile` contained `PATH="/usr/local/sbin:/usr/local/bin:/usr/bin"` instead of this `appendpath` function that' used now:

https://github.com/archlinux/svntogit-packages/commit/20928f58767d34ed6711befd6255f6a0b1706ae8#diff-0fd4044803ddc7f479169e5a143e81b5

So with that updated `/etc/profile`, loading `.profile` from `.zshrc` is no longer necessary (except maybe for the order of the paths?). Not sure if we can assume everyone using this config has a `/etc/profile` that doesn't just overwrite `PATH`.

**Comment by m-ou-se** [comment]
The behaviour of `zsh` is still the same as a few years ago. It loads `/etc/profile` after `zshenv` but before `zshrc`:

```
$ zsh -o sourcetrace -l
+/home/mara/.zshenv:1> <sourcetrace>
+/home/mara/.profile:1> <sourcetrace>
+/etc/zsh/zprofile:1> <sourcetrace>
+/etc/profile:1> <sourcetrace>
(...)
+/home/mara/.zshrc:1> <sourcetrace>
+/home/mara/.profile:1> <sourcetrace>
(...)
```

**Comment by Minnozz** [comment]
Interesting, on my machine (Ubuntu 20.04), zsh does not source `/etc/profile` at all:
```
$ zsh -o sourcetrace -l
+/etc/zsh/zshenv:1> <sourcetrace>
+/etc/zsh/zprofile:1> <sourcetrace>
+/home/minnozz/.zprofile:1> <sourcetrace>
+/home/minnozz/.config/config_dir/profile:1> <sourcetrace>
+/etc/zsh/zshrc:1> <sourcetrace>
+/home/minnozz/.zcompdump:1> <sourcetrace>
+/home/minnozz/.zshrc:1> <sourcetrace>
(...)
```
(I use a slightly different config setup, inspired by this repository)

My `/etc/zsh/zprofile` contains only comments. Does yours maybe explicitly source `/etc/profile`?

**Comment by m-ou-se** [comment]
> Does yours maybe explicitly source /etc/profile?

Yes.

**Comment by de-vri-es** [comment]
Right, we made `.profile` idempotent partly so we could source it twice without problem. But it is also required to deal with a shell started from within a shell. Sourcing `.profile` twice does mean that `.profile.local` needs to be idempotent, but since that's kinda needed anyway I don't really see it as a problem.

Note that on Arch the order of `$PATH` should be fine regardless of the order of sourcing `/etc/profile` and `$HOME/.profile`. Our `add_path` function prepends, while `/etc/profile` appends.

If a system has an `/etc/profile` that prepends to `$PATH`, we would currently end up with the wrong order. Since `$HOME/.zshenv` is sourced before `/etc/profile`, the paths from `/etc/profile` would end up before ours. That would happen even in non-interactive shells that are started from an interactive shell, since it would inherit `$PATH`).

So I'm not really sure if there is anything to be done here. I suppose it would be nice if we could work with `/etc/profile` that prepends. But that does feel like `/etc/profile` would be doing it wrong, and I don't know if there are any distros that actually do that by default.
